src/librustc_typeck/check/mod.rs

```diff
@@ -1353,9 +1353,41 @@ fn check_union(tcx: TyCtxt<'_>, id: hir::HirId, span: Span) {
     def.destructor(tcx); // force the destructor to be evaluated
     check_representable(tcx, span, def_id);
     check_transparent(tcx, span, def_id);
+    check_union_fields(tcx, span, def_id);
     check_packed(tcx, span, def_id);
 }
 
+/// When the `#![feature(untagged_unions)]` gate is active,
+/// check that the fields of the `union` does not contain fields that need dropping.
+fn check_union_fields(tcx: TyCtxt<'_>, _: Span, item_def_id: DefId) -> bool {
+    // Without the feature we check that all fields are `Copy` in our stability checking
+    // infrastructure.
```

Comment:
Please mention a filename at least. I have no idea where the "stability checking infrastructure" lives.

Also, why does it make sense to split the check for union fields across two entirely separate passes? Seems like spaghetti code to me.

Reply:
Someone else is of the same opinion: https://github.com/rust-lang/rust/blob/b4748679611c97c09cfb7096917d7830c410a4ad/src/librustc/middle/stability.rs#L832-L833 but couldn't find a better solution it seems

Reply:
Well, having that in stability kind of makes sense I guess as it is about a feature gate.

What happens if you remove the `tcx.features().untagged_unions` check here? That would at least remove the danger of forgetting to remove this later.

Reply:
How could we forget to remove it? The `untagged_unions` field should disappear when we stabilize the feature and even if not, we'd just get errors in our tests.

Reply:
If we removed the check, we'd still check for copy types even with the feature gate active

Reply:
> If we removed the check, we'd still check for copy types even with the feature gate active

Now I am confused. I am talking about removing the check to which this GH discussion is attached. If we remove that, we'd *always* do the `needs_drop` check *even if* the feature gate is disabled -- right?

Reply:
I inverted the condition in my head. It's fixed now